thirdparty/DiskANN/src/pq_flash_index.cpp

```diff
@@ -1566,28 +1557,46 @@ namespace diskann {
     auto ctx = this->reader->get_ctx();
     const auto sector_num = sector_offsets.size();
     const _u64 num_blocks = DIV_ROUND_UP(sector_num, batch_size);
+    std::vector<AlignedRead> last_reqs;
+    bool rotate = false;
+
     for (_u64 i = 0; i < num_blocks; ++i) {
       _u64 start_idx = i * batch_size;
       _u64 idx_len = std::min(batch_size, sector_num - start_idx);
+      last_reqs = frontier_read_reqs;
       frontier_read_reqs.clear();
       for (_u64 j = 0; j < idx_len; ++j) {
-        char *sector_buf = sector_scratch + j * read_len_for_node;
+        char *sector_buf =
+            sector_scratch + rotate * half_buf_idx + j * read_len_for_node;
         frontier_read_reqs.emplace_back(sector_offsets[start_idx + j], read_len_for_node,
                                     sector_buf);
       }
-      reader->read(frontier_read_reqs, ctx);
-      for (const auto& req : frontier_read_reqs) {
+      rotate ^= 0x1;
+      reader->submit_req(ctx, frontier_read_reqs);
+      for (const auto& req : last_reqs) {
         auto offset = req.offset;
         char* sector_buf = static_cast<char*>(req.buf);
-        for (auto idx : sectors_to_visit[offset]) {
+        for (auto idx : sectors_to_visit.at(offset)) {
           char* node_buf = get_offset_to_node(sector_buf, ids[idx]);
           copy_vec_base_data(output_data, idx, node_buf);
         }
       }
+      reader->get_submitted_req(ctx, frontier_read_reqs.size());
```

Comment:
Looks like the pipeline is:
submit 1->copy out 0->get 1->submit 2->copy out 1->get 2;
Did we try:
submit 1->get 0->copy out 0->submit2->get 1->copy out1;
So we can have one more operation between submit/get.
Just brainstorming, I don't know specific bottleneck of this function

Reply:
The latter approach involves two contexts and related events instead of one, the number of contexts is limited in the system. This may affect throughput and latency, to be tested.

Reply:
Do you want to have this experimented before merging this PR?

Reply:
Updated in the PR description section